core/src/main/scala/org/apache/spark/Partitioner.scala

```diff
@@ -118,7 +118,8 @@ class RangePartitioner[K : Ordering : ClassTag, V](
       Array.empty
     } else {
       // This is the sample size we need to have roughly balanced output partitions, capped at 1M.
-      val sampleSize = math.min(20.0 * partitions, 1e6)
+      val maxSamples = rdd.sparkContext.getConf.getDouble("spark.partitioner.max_samples", 1e6)
```

Comment:
Not sure how 1e6 was arrived at, but for our jobs at anything > 40k partitions, for non primitive keys we hit the issue. The actual number of keys is not too large, which is why lower the value does not affect precision of the estimate.